File: typescript/src/core/compose-bindings.ts
import type { Place } from './place.js';
import type { Transition } from './transition.js';

/**
 * @internal Symbol key restricting construction to
 * {@link PetriNetBuilder.compose}. Bindings are produced by the host builder
 * and supplied to the caller's callback — direct construction is not part of
 * the public API.
 */
const COMPOSE_BINDINGS_KEY = Symbol('ComposeBindings.internal');

/**
 * Typed binding builder for {@link PetriNetBuilder.compose}, per
 * `spec/11-modular-composition.md` requirements **MOD-020** (composition
 * operation) and **MOD-022** (type compatibility).
 *
 * `ComposeBindings` is a write-only collector: callers register port and
 * channel bindings against an instance's interface, and the host builder
 * consumes the recorded mappings as it merges the instance into the
 * enclosing net.
 *
 * ## Port bindings (MOD-020, MOD-022)
 *
 * {@link bindPort} merges an interface port with a caller-side place. The
 * port name is the **original** (pre-prefix) name declared in the subnet's
 * `Interface`; the caller place takes the port's slot in the resulting net.
 * Token-type compatibility is enforced at compile time only in TypeScript
 * (per [MOD-022]) — the typed `bindPort<T>` signature is the safety
 * mechanism. TypeScript erases generics at runtime, so a misuse via `as`
 * casts is structurally undetectable.
 *
 * ## Channel bindings (MOD-021)
 *
 * {@link bindChannel} records a synchronous channel binding: at compose time
 * the named instance-side interface transition is **merged** with the
 * supplied caller-side {@link Transition} into one transition in the
 * resulting flat net. Channel composition is implemented in task #13; until
 * then `compose(...)` raises an `Error` when any channel binding is present.
 *
 * ## Identity
 *
 * Instances are produced by {@link PetriNetBuilder.compose} and supplied to
 * the caller's callback. The constructor is symbol-guarded to prevent
 * direct construction.
 */
export class ComposeBindings {
  private readonly _portBindings = new Map<string, Place<unknown>>();
  private readonly _channelBindings = new Map<string, Transition>();

  /**
   * @internal Use {@link PetriNetBuilder.compose} — instances are produced
   * by the host builder and supplied to the caller's callback.
   */
  constructor(key: symbol) {
    if (key !== COMPOSE_BINDINGS_KEY) {
      throw new Error(
        'Use PetriNetBuilder.compose(instance, b => ...) — ComposeBindings is not directly constructible',
      );
    }
  }

  /**
   * Binds the named interface port to the given caller place per **MOD-020**.
   *
   * The port name is the **original** (pre-prefix) name declared in the
   * subnet's `Interface`. The typed `<T>` parameter ensures the caller
   * place's token type matches the interface port's token type at compile
   * time per [MOD-022]; TypeScript does not validate the type at runtime
   * because generics are erased.
   *
   * @throws when `portName` is already bound on this builder
   */
  bindPort<T>(portName: string, callerPlace: Place<T>): this {
    if (this._portBindings.has(portName)) {
      throw new Error(`Port '${portName}' is already bound`);
    }
    this._portBindings.set(portName, callerPlace as Place<unknown>);
    return this;
  }

  /**
   * Records a synchronous channel binding per **MOD-021**: at compose time,
   * the instance-side renamed channel transition is merged with
   * `callerTransition` into a single transition in the resulting flat net.
   *
   * **Status**: channel composition is implemented in task #13. Recording a
   * channel binding here is permitted, but `compose(...)` currently raises
   * an `Error` when any channel binding is present.
   *
   * @throws when `channelName` is already bound on this builder
   */
  bindChannel(channelName: string, callerTransition: Transition): this {
    if (this._channelBindings.has(channelName)) {
      throw new Error(`Channel '${channelName}' is already bound`);
    }
    this._channelBindings.set(channelName, callerTransition);
    return this;
  }

  /** Returns an unmodifiable view of the recorded port bindings. */
  portBindings(): ReadonlyMap<string, Place<unknown>> {
    return this._portBindings;
  }

  /** Returns an unmodifiable view of the recorded channel bindings. */
  channelBindings(): ReadonlyMap<string, Transition> {
    return this._channelBindings;
  }
}

/**
 * @internal Package-internal factory used by {@link PetriNetBuilder.compose}
 * to construct {@link ComposeBindings} values without re-exporting the
 * Symbol-guarded constructor key publicly.
 *
 * NOT part of the public API surface — do NOT re-export from `core/index.ts`.
 */
export function __createComposeBindings(): ComposeBindings {
  return new ComposeBindings(COMPOSE_BINDINGS_KEY);
}

File: java/src/main/java/org/libpetri/core/ComposeBindings.java
package org.libpetri.core;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

/**
 * Typed binding builder for {@link PetriNet.Builder#compose(Instance, java.util.function.Consumer)},
 * per {@code spec/11-modular-composition.md} requirements <b>MOD-020</b>
 * (composition operation) and <b>MOD-022</b> (type compatibility).
 *
 * <p>{@code ComposeBindings} is a write-only collector: callers register port
 * and channel bindings against an instance's interface, and the host
 * builder consumes the recorded mappings as it merges the instance into the
 * enclosing net.
 *
 * <h2>Port bindings (MOD-020, MOD-022)</h2>
 * <p>{@link #bindPort(String, Place)} merges an interface port with a
 * caller-side place. The port name is the <b>original</b> (pre-prefix) name
 * declared in the subnet's {@link Interface}; the caller place takes the
 * port's slot in the resulting net. Token-type compatibility is enforced at
 * compose time per [MOD-022].
 *
 * <h2>Channel bindings (MOD-021)</h2>
 * <p>{@link #bindChannel(String, Transition)} records a synchronous channel
 * binding: at compose time, the named instance-side interface transition is
 * <b>merged</b> with the supplied caller-side {@link Transition} into one
 * transition in the resulting flat net. Arc union, sequential action
 * composition, and timing/priority resolution follow the merge contract per
 * [MOD-021]; see {@link PetriNet.Builder#compose(Instance, java.util.function.Consumer)}.
 *
 * <h2>Identity</h2>
 * <p>{@code ComposeBindings} is a final class — not a record, not extensible.
 * It carries mutable state during the {@code Consumer<ComposeBindings>}
 * callback and is then drained by the compose call.
 */
public final class ComposeBindings {

    private final LinkedHashMap<String, Place<?>> portBindings = new LinkedHashMap<>();
    private final LinkedHashMap<String, Transition> channelBindings = new LinkedHashMap<>();

    /**
     * Package-private constructor — instances are produced by
     * {@link PetriNet.Builder#compose(Instance, java.util.function.Consumer)}
     * and supplied to the caller's {@link java.util.function.Consumer}.
     */
    ComposeBindings() { }

    /**
     * Binds the named interface port to the given caller place per <b>MOD-020</b>.
     *
     * <p>The port name is the <b>original</b> (pre-prefix) name declared in
     * the subnet's {@link Interface}. The caller place's token type must
     * match the interface place's token type per <b>MOD-022</b> — this is
     * verified at compose time, not at {@code bindPort} time, because the
     * lookup needs the resolved {@link Instance}.
     *
     * @param portName     the port name as declared in the subnet's {@link Interface}
     * @param callerPlace  the caller-side place to merge with the interface port
     * @param <T>          token type of the caller place
     * @return this binding builder, for chaining
     * @throws IllegalArgumentException when {@code portName} or {@code callerPlace}
     *         is null, or when {@code portName} was already bound
     */
    public <T> ComposeBindings bindPort(String portName, Place<T> callerPlace) {
        Objects.requireNonNull(portName, "portName");
        Objects.requireNonNull(callerPlace, "callerPlace");
        if (portBindings.containsKey(portName)) {
            throw new IllegalArgumentException(
                "Port '" + portName + "' is already bound");
        }
        portBindings.put(portName, callerPlace);
        return this;
    }

    /**
     * Records a synchronous channel binding per <b>MOD-021</b>: at compose
     * time, the instance-side renamed channel transition (resolved by name on
     * the {@link Instance}) is merged with {@code callerTransition} into a
     * single transition in the resulting flat net.
     *
     * <p>The arc set, action, timing, and priority of the merged transition
     * are computed per {@link PetriNet.Builder#compose(Instance, java.util.function.Consumer)}.
     * The caller-side transition's name and identity become the merged
     * transition's name and identity (caller-wins convention).
     *
     * @param channelName        the channel name as declared in the subnet's {@link Interface}
     * @param callerTransition   the caller-side transition to fuse with the interface channel
     * @return this binding builder, for chaining
     * @throws IllegalArgumentException when {@code channelName} or
     *         {@code callerTransition} is null, or when {@code channelName}
     *         was already bound
     */
    public ComposeBindings bindChannel(String channelName, Transition callerTransition) {
        Objects.requireNonNull(channelName, "channelName");
        Objects.requireNonNull(callerTransition, "callerTransition");
        if (channelBindings.containsKey(channelName)) {
            throw new IllegalArgumentException(
                "Channel '" + channelName + "' is already bound");
        }
        channelBindings.put(channelName, callerTransition);
        return this;
    }

    /** Returns an unmodifiable view of the recorded port bindings. */
    public Map<String, Place<?>> portBindings() {
        return Collections.unmodifiableMap(portBindings);
    }

    /** Returns an unmodifiable view of the recorded channel bindings. */
    public Map<String, Transition> channelBindings() {
        return Collections.unmodifiableMap(channelBindings);
    }
}
